src/container/srv_target.c

```diff
@@ -2378,131 +2396,229 @@ ds_cont_oid_alloc_handler(crt_rpc_t *rpc)
-	return 0;
+		DP_UUID(cont_uuid), tgt_idx, ec_eph->cte_ref);
+	return;
 }
 
 static void
-cont_ec_eph_destroy(struct cont_ec_eph *ec_eph)
+cont_track_eph_destroy(struct cont_track_eph *ec_eph)
 {
-	D_ASSERT(ec_eph->ce_ref == 0);
-	d_list_del(&ec_eph->ce_list);
-	D_FREE(ec_eph->ce_ephs);
+	D_ASSERT(ec_eph->cte_ref == 0);
+	d_list_del(&ec_eph->cte_list);
+	D_FREE(ec_eph->cte_stable_ephs);
+	D_FREE(ec_eph->cte_ec_agg_ephs);
 	D_FREE(ec_eph);
 }
 
 void
-ds_cont_ec_eph_free(struct ds_pool *pool)
+ds_cont_track_eph_free(struct ds_pool *pool)
+{
+	struct cont_track_eph	*ec_eph, *tmp;
+
+	d_list_for_each_entry_safe(ec_eph, tmp, &pool->sp_ec_ephs_list, cte_list)
+		cont_track_eph_destroy(ec_eph);
+}
+
+struct track_eph_ult_arg {
+	struct ds_pool	*pool;
+	uuid_t		cont_uuid;
+	uint32_t	tgt_idx;
+	daos_epoch_t	*ec_agg_eph;
+	daos_epoch_t	*stable_eph;
+};
+
+static	int
+cont_track_eph_fini_ult(void *data)
+{
+	struct track_eph_ult_arg	*arg = data;
+
+	cont_track_eph_delete(arg->pool, arg->cont_uuid, arg->tgt_idx);
+	return 0;
+}
+
+static void
+cont_tgt_track_eph_fini(struct ds_cont_child *cont_child)
+{
+	struct track_eph_ult_arg	arg;
+
+	if (cont_child->sc_query_ec_agg_eph == NULL)
+		return;
+	D_ASSERT(cont_child->sc_query_stable_eph != NULL);
+
+	arg.pool = cont_child->sc_pool->spc_pool;
+	uuid_copy(arg.cont_uuid, cont_child->sc_uuid);
+	arg.tgt_idx = dss_get_module_info()->dmi_tgt_id;
+	dss_ult_execute(cont_track_eph_fini_ult, &arg, NULL, NULL, DSS_XS_SYS, 0, 0);
+
+	cont_child->sc_query_ec_agg_eph = NULL;
+	cont_child->sc_query_stable_eph = NULL;
+}
+
+static int
+cont_track_eph_init_ult(void *data)
 {
-	struct cont_ec_eph	*ec_eph, *tmp;
+	struct track_eph_ult_arg *arg = data;
+	int rc;
 
-	d_list_for_each_entry_safe(ec_eph, tmp, &pool->sp_ec_ephs_list, ce_list)
-		cont_ec_eph_destroy(ec_eph);
+	rc = cont_track_eph_insert(arg->pool, arg->cont_uuid, arg->tgt_idx, &arg->ec_agg_eph,
+				   &arg->stable_eph);
+	return rc;
+}
+
+static void
+cont_tgt_track_eph_init_ult(void *data)
+{
+	struct ds_cont_child		*cont_child = data;
+	struct track_eph_ult_arg	arg;
+	int				rc;
+
+	arg.pool = cont_child->sc_pool->spc_pool;
+	uuid_copy(arg.cont_uuid, cont_child->sc_uuid);
+	arg.tgt_idx = dss_get_module_info()->dmi_tgt_id;
+	rc = dss_ult_execute(cont_track_eph_init_ult, &arg, NULL, NULL, DSS_XS_SYS,
+			     0, 0);
+	if (rc) {
+		DL_ERROR(rc, DF_CONT " init track eph failed.\n",
+			 DP_CONT(cont_child->sc_pool->spc_uuid, cont_child->sc_uuid));
+		ds_cont_child_put(cont_child);
+		return;
+	}
+
+	D_DEBUG(DB_MD, DF_UUID " update init track %u\n",
+		DP_UUID(cont_child->sc_uuid), arg.tgt_idx);
+	cont_child->sc_query_ec_agg_eph = arg.ec_agg_eph;
+	cont_child->sc_query_stable_eph = arg.stable_eph;
+
+	ds_cont_child_put(cont_child);
+}
+
+static int
+cont_tgt_track_eph_init(struct ds_cont_child *cont_child)
+{
+	int rc;
+
+	ds_cont_child_get(cont_child);
+
+	rc = dss_ult_create(cont_tgt_track_eph_init_ult, cont_child, DSS_XS_SELF,
```

Comment:
cont_tgt_track_eph_init_ult() will create another ULT cont_track_eph_init_ult(), why need two ULTs for that?

Reply:
cont_tgt_track_eph_init_ult() execute on vos tgt main XS, it will execute cont_track_eph_init_ult on XS0. The pool->sp_ec_ephs_list only maintained (insert/lookup) on XS0

Reply:
the pool->sp_ec_ephs_list only maintained on XS0, so create ULT and execute on XS0 need not take lock.
hmm, maybe need not create cont_tgt_track_eph_init_ult, just create cont_track_eph_init_ult on XS0, is this what you prefer?

Reply:
Yes, I just don't see why we need to create a ULT for creating a ULT on sys xstream. This dss_ult_create() looks not necessary to me, it also makes cont_track_eph_insert() not being executed before cont_tgt_track_eph_init() returns, that looks not quite right to me.

Reply:
OK, i see, will change it. thx